# Reject malformed node and element input
#
#   - elements: reject nonpositive element numbers and undefined nodes
#   - boundarys: reject negative degrees of freedom
#   - expansions, electricalconductivitys: check nmat before using it
#   - nodalthicknesss: require 1-D/2-D elements and defined nodes
#   - pretensionsections: require the NODE parameter
#   - refinemeshs: require a positive LIMIT
#   - near3d: stop the search once all nodes have been visited
#   - calinput: reject massless contact outside explicit dynamics
#
# Found by fuzzing ccx with malformed variants of 518 input decks under
# AddressSanitizer, and by review of the code around each crash.

diff --git a/ccx_2.23/src/boundarys.f b/ccx_2.23/src/boundarys.f
index 9d75cbc..70e84f6 100644
--- a/ccx_2.23/src/boundarys.f
+++ b/ccx_2.23/src/boundarys.f
@@ -348,6 +348,14 @@
               return
             endif
           endif
+!     
+          if((ibounstart.lt.0).or.(ibounend.lt.0)) then
+            write(*,*) '*ERROR reading *BOUNDARY: nonexistent'
+            write(*,*) '       degree of freedom'
+            call inputerror(inpc,ipoinpc,iline,
+     &           "*BOUNDARY%",ier)
+            return
+          endif
 !     
           if(textpart(4)(1:1).eq.' ') then
             bounval=0.d0
diff --git a/ccx_2.23/src/calinput.f b/ccx_2.23/src/calinput.f
index 0402eba..f27e88f 100644
--- a/ccx_2.23/src/calinput.f
+++ b/ccx_2.23/src/calinput.f
@@ -645,7 +645,8 @@ c
      &       ne,ne_,set,istartset,iendset,ialset,nset,nset_,nalset,
      &       nalset_,mi(1),ixfree,iponor,xnor,istep,istat,n,iline,
      &       ipol,inl,ipoinp,inp,iaxial,ipoinpc,solid,
-     &       network,filab,nlabel,out3d,iuel,nuel_,ier,iparentel)
+     &       network,filab,nlabel,out3d,iuel,nuel_,ier,iparentel,
+     &       nk)
 !     
       elseif((textpart(1)(1:7).eq.'*ELFILE').or.
      &       (textpart(1)(1:14).eq.'*ELEMENTOUTPUT')) then
@@ -897,7 +898,7 @@ c
       elseif(textpart(1)(1:15).eq.'*NODALTHICKNESS') then
         call nodalthicknesss(inpc,textpart,set,istartset,iendset,
      &       ialset,nset,thickn,nk,istep,istat,n,iline,ipol,inl,
-     &       ipoinp,inp,iaxial,ipoinpc,ier)
+     &       ipoinp,inp,iaxial,ipoinpc,ier,ne1d,ne2d)
 !     
       elseif((textpart(1)(1:5).eq.'*NODE').and.
      &       (textpart(1)(1:10).ne.'*NODEPRINT').and.
@@ -1259,6 +1260,16 @@ c
       endif
 !     
       enddo loop
+!     
+!     massless contact is only allowed in explicit dynamic steps
+!     
+      if((istep.gt.0).and.(mortar.eq.-1).and.
+     &     ((nmethod.ne.4).or.(iexpl.le.1))) then
+        write(*,*) '*ERROR in calinput: massless contact is only'
+        write(*,*) '       allowed in an explicit dynamic step'
+        write(*,*) '       (*DYNAMIC,EXPLICIT)'
+        ier=1
+      endif
 !     
       if(ier.ge.1) then
         write(*,*) '*ERROR in calinput: at least one fatal'
diff --git a/ccx_2.23/src/calinput_rfn.f b/ccx_2.23/src/calinput_rfn.f
index a84f67d..5927870 100644
--- a/ccx_2.23/src/calinput_rfn.f
+++ b/ccx_2.23/src/calinput_rfn.f
@@ -66,7 +66,8 @@
      &         ne,ne_,set,istartset,iendset,ialset,nset,nset_,nalset,
      &         nalset_,mi(1),ixfree,iponor,xnor,istep,istatrfn,n,iline,
      &         ipol,inl,ipoinp,inp,iaxial,ipoinpc,solid,
-     &         network,filab,nlabel,out3d,iuel,nuel_,ier,iparentel)
+     &         network,filab,nlabel,out3d,iuel,nuel_,ier,iparentel,
+     &         nk)
 !     
         elseif(textpart(1)(1:12).eq.'*MODELCHANGE') then
           call modelchanges(inpc,textpart,tieset,istatrfn,n,iline,
diff --git a/ccx_2.23/src/electricalconductivitys.f b/ccx_2.23/src/electricalconductivitys.f
index 837151e..2da15c4 100644
--- a/ccx_2.23/src/electricalconductivitys.f
+++ b/ccx_2.23/src/electricalconductivitys.f
@@ -34,7 +34,6 @@
       real*8 alcon(0:6,ntmat_,*),alzero(*)
 !
       ntmat=0
-      alzero(nmat)=0.d0
 !
       if((istep.gt.0).and.(irstrt(1).ge.0)) then
          write(*,*) '*ERROR reading *ELECTRICAL CONDUCTIVITY:'
@@ -51,6 +50,8 @@
          ier=1
          return
       endif
+!
+      alzero(nmat)=0.d0
 !
       ityp=1
 !
diff --git a/ccx_2.23/src/elements.f b/ccx_2.23/src/elements.f
index 6f99c63..212122d 100644
--- a/ccx_2.23/src/elements.f
+++ b/ccx_2.23/src/elements.f
@@ -20,7 +20,7 @@
      &     set,istartset,iendset,ialset,nset,nset_,nalset,nalset_,mi,
      &     ixfree,iponor,xnor,istep,istat,n,iline,ipol,inl,ipoinp,inp,
      &     iaxial,ipoinpc,solid,network,filab,nlabel,out3d,iuel,
-     &     nuel_,ier,iparentel)
+     &     nuel_,ier,iparentel,nk)
 !     
 !     reading the input deck: *ELEMENT
 !     
@@ -39,7 +39,8 @@
      &     nteller,j,ipkon(*),nkon,nope,indexe,mi(*),ipos,indexy,ixfree,
      &     iponor(2,*),nopeexp,iline,ipol,inl,ipoinp(2,*),inp(3,*),
      &     iaxial,ipoinpc(0:*),nlabel,network,iuel(4,*),nuel_,
-     &     id,four,number,ndof,intpoints,ier,iparentel(*),iparent
+     &     id,four,number,ndof,intpoints,ier,iparentel(*),iparent,
+     &     nk,node
 !     
       real*8 xnor(*)
 !     
@@ -434,6 +435,13 @@ c     Bernhardi end
         read(textpart(1)(1:10),'(i10)',iostat=istat) i
         if(istat.gt.0) then
           call inputerror(inpc,ipoinpc,iline,
+     &         "*ELEMENT%",ier)
+          return
+        endif
+        if(i.le.0) then
+          write(*,*) '*ERROR reading *ELEMENT: element number',i
+          write(*,*) '       must be positive'
+          call inputerror(inpc,ipoinpc,iline,
      &         "*ELEMENT%",ier)
           return
         endif
@@ -496,6 +504,22 @@ c     Bernhardi end
             if(nteller.eq.nope) exit
           enddo
         endif
+!     
+!     check whether the nodes of the element are defined; node 0
+!     denotes the entry or exit of a network element
+!     
+        do j=1,nope
+          node=kon(indexe+j)
+          if((node.gt.0).and.(node.le.nk)) cycle
+          if((node.eq.0).and.(j.ne.2).and.(label(1:1).eq.'D').and.
+     &         (label(1:7).ne.'DCOUP3D')) cycle
+          write(*,*) '*ERROR reading *ELEMENT: node',node
+          write(*,*) '       of element',i,' is not defined'
+          call inputerror(inpc,ipoinpc,iline,
+     &         "*ELEMENT%",ier)
+          return
+        enddo
+!     
         ne=max(ne,i)
 !     
         if(iparent.gt.0) iparentel(i)=iparent
diff --git a/ccx_2.23/src/expansions.f b/ccx_2.23/src/expansions.f
index eb20728..063ea63 100644
--- a/ccx_2.23/src/expansions.f
+++ b/ccx_2.23/src/expansions.f
@@ -34,7 +34,6 @@
       real*8 alcon(0:6,ntmat_,*),alzero(*)
 !
       ntmat=0
-      alzero(nmat)=0.d0
 !
       if((istep.gt.0).and.(irstrt(1).ge.0)) then
          write(*,*) 
@@ -51,6 +50,8 @@
          ier=1
          return
       endif
+!
+      alzero(nmat)=0.d0
 !
       ityp=1
 !
diff --git a/ccx_2.23/src/near3d.f b/ccx_2.23/src/near3d.f
index 82e9112..4304ec7 100644
--- a/ccx_2.23/src/near3d.f
+++ b/ccx_2.23/src/near3d.f
@@ -212,6 +212,11 @@ c         call dsort(c,idummy,eight,iflag)
             endif
          endif
          if(c(1).ge.r(k)) exit
+!
+!        all nodes have been checked (the test above never succeeds
+!        if r(k) is NaN or infinite, e.g. for diverged coordinates)
+!
+         if(i.gt.n) exit
 !
          i=i+1
 !
diff --git a/ccx_2.23/src/nodalthicknesss.f b/ccx_2.23/src/nodalthicknesss.f
index 396ce34..19d7572 100644
--- a/ccx_2.23/src/nodalthicknesss.f
+++ b/ccx_2.23/src/nodalthicknesss.f
@@ -18,7 +18,7 @@
 !
       subroutine nodalthicknesss(inpc,textpart,set,istartset,iendset,
      &  ialset,nset,thickn,nk,istep,istat,n,iline,ipol,inl,ipoinp,
-     &  inp,iaxial,ipoinpc,ier)
+     &  inp,iaxial,ipoinpc,ier,ne1d,ne2d)
 !
 !     reading the input deck: *NODAL THICKNESS
 !
@@ -30,7 +30,7 @@
 !
       integer istartset(*),iendset(*),ialset(*),nset,nk,istep,istat,n,
      &  key,i,j,k,l,ipos,iline,ipol,inl,ipoinp(2,*),inp(3,*),iaxial,
-     &  ipoinpc(0:*),ier,id
+     &  ipoinpc(0:*),ier,id,ne1d,ne2d
 !
       real*8 thickn(2,*),thickness1,thickness2
 !
@@ -40,6 +40,17 @@
          ier=1
          return
       endif
+!
+!     the nodal thickness field only exists if 1-D or 2-D elements
+!     were defined
+!
+      if((ne1d.eq.0).and.(ne2d.eq.0)) then
+         write(*,*) '*ERROR reading *NODAL THICKNESS: no 1-D or 2-D'
+         write(*,*) '       elements were defined'
+         call inputerror(inpc,ipoinpc,iline,
+     &        "*NODAL THICKNESS%",ier)
+         return
+      endif
 !
       do i=2,n
          write(*,*) 
@@ -73,6 +84,12 @@
          endif
          read(textpart(1)(1:10),'(i10)',iostat=istat) l
          if(istat.eq.0) then
+            if((l.gt.nk).or.(l.le.0)) then
+               write(*,*) '*ERROR reading *NODAL THICKNESS:'
+               write(*,*) '       node ',l,' is not defined'
+               ier=1
+               return
+            endif
             thickn(1,l)=thickness1
             thickn(2,l)=thickness2
          else
diff --git a/ccx_2.23/src/pretensionsections.f b/ccx_2.23/src/pretensionsections.f
index d76a1a3..e372d06 100644
--- a/ccx_2.23/src/pretensionsections.f
+++ b/ccx_2.23/src/pretensionsections.f
@@ -113,6 +113,7 @@
       endif
 !     
       ielem=0
+      irefnode=0
       do i=1,81
         surface(i:i)=' '
       enddo
@@ -178,6 +179,17 @@
         endif
       enddo
 !     
+!     checking whether a reference node was specified
+!     
+      if(irefnode.eq.0) then
+        write(*,*) '*ERROR reading *PRE-TENSION SECTION:'
+        write(*,*) '       no reference node (parameter NODE)'
+        write(*,*) '       specified'
+        call inputerror(inpc,ipoinpc,iline,
+     &       "*PRE-TENSION SECTION%",ier)
+        return
+      endif
+!     
 !     checking whether the surface exists and is an element face
 !     surface
 !     
diff --git a/ccx_2.23/src/refinemeshs.f b/ccx_2.23/src/refinemeshs.f
index a4f5bbb..3dfcb1a 100644
--- a/ccx_2.23/src/refinemeshs.f
+++ b/ccx_2.23/src/refinemeshs.f
@@ -32,6 +32,8 @@
 !     
       integer istep,istat,n,key,i,ii,ier,iline,ipol,inl,ipoinp(2,*),
      &     inp(3,*),ipoinpc(0:*),nset,ipos,id
+!
+      real*8 targetsize
 !     
       if(istep.lt.1) then
         write(*,*) 
@@ -102,6 +104,20 @@
           filab(48)(i:i)=' '
         enddo
       endif
+!     
+!     the target size (parameter LIMIT) must be positive unless
+!     only smoothing is requested
+!     
+      if(.not.smoothingonly) then
+        read(filab(48)(7:26),'(f20.0)',iostat=istat) targetsize
+        if((istat.gt.0).or.(targetsize.le.0.d0)) then
+          write(*,*) '*ERROR reading *REFINE MESH: parameter LIMIT'
+          write(*,*) '       is missing or not positive'
+          call inputerror(inpc,ipoinpc,iline,
+     &         "*REFINE MESH%",ier)
+          return
+        endif
+      endif
 !     
       if(smoothingonly) then
         filab(48)(3:6)='SMOO'
